### PR TITLE
Introduce Use Case Preprocessor

### LINKED ISSUE #426: Simplify Defining Use Case Tests
## Description ##
As a SE, I would like to have an easy-to-use tool for use case definition, so I can assure simplicity, completeness, and good functional coverage for use-case suite.

## Acceptance Criteria ##
- Provide the possibility to specify the expected result of the analysis without manual counting of positions
- Cover all the possible cases - errors, variables, paragraphs, copybooks, with definition, and replacing
- Rewrite the already existing tests when it is possible

## Definition of Done ##
- The story is implemented according to the description.
- Unit testing done (Preferably by TDD approach, coverage should be at least 70%).
- Code review done by at least one other team member.
- Manual testing is done.
- Documentation is added and reviewed (if needed).
- The pull request is approved.
- The story is merged into the development branch.
- Tests are performed on the development branch.

## Technical details ##
This is the place for any implementation-related notes

### PR DESCRIPTION
**Pull request by temanbrcom**
An easy-to-use tool for use case definition to assure simplicity, completeness, and good functional coverage for use-case suite.
Closes #426

### CODE REVIEW

**Comment by temanbrcom** [comment]
@grianbrcom 
@asatklichov

File: com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/delegates/validations/SourceInfoLevels.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * This class contains human-readable text describing the source of diagnostics, including severity
+ * level.
+ */
+@AllArgsConstructor
+public enum SourceInfoLevels {
+  ERROR("E"),
+  WARNING("W"),
+  INFO("I"),
+  HINT("I");
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/delegates/validations/SourceInfoLevels.java:29]: "I", I guess point is "Hint" is "Info" at the end, right?

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/service/delegates/validations/UseCaseUtils.java [hunk 2]
```diff
@@ -79,9 +85,13 @@ public static void runTextValidation(TextDocumentService service, String text) {
    * implementation uses the default time to await. Checker example: await( () -> { return
    * !client.getReceivedDiagnostics().isEmpty(); });
    *
+   * <p>Deprecated - try using analyze, analyzeForErrors, or {@link
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/service/delegates/validations/UseCaseUtils.java:88]: In comment use @deprecated

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/service/delegates/validations/UseCaseUtils.java [hunk 3]
```diff
@@ -92,10 +102,14 @@ public static void await(Callable<Boolean> checker) {
...
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/service/delegates/validations/UseCaseUtils.java:105]: same

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/service/delegates/validations/UseCaseUtils.java [hunk 4]
```diff
@@ -105,11 +119,15 @@ private static void await(Callable<Boolean> checker, String description) {
    * occurred. WARNING: use only in the thread that has been used to run the server. Checker
    * example: await( () -> { return !client.getReceivedDiagnostics().isEmpty(); });
    *
+   * <p>Deprecated - try using analyze, analyzeForErrors, or {@link
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/service/delegates/validations/UseCaseUtils.java:122]: @deprecated

File: com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/delegates/validations/SourceInfoLevels.java (hunk 1 above)
Comment (temanbrcom) [reply com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/delegates/validations/SourceInfoLevels.java:29]: Oh, it's a typo, my bad. 
The order is Error > Warning > Info > Hint

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCopybookWithIndirectRecursiveDependencyIsDetected.java [hunk 5]
```diff
@@ -15,54 +15,63 @@
...
   @Test
   public void test() {
-    super.test(
-        List.of(
-            new CobolText("INNER-COPY", INNER_COPY),
-            new CobolText("INDIRECT-COPY", INDIRECT_COPY)));
-  }
+    Map<String, Diagnostic> expectedDiagnostics = new HashMap<>();
+    expectedDiagnostics.put(
+        "1", new Diagnostic(null, MESSAGE_RECURSION + INDIRECT_NAME, Error, ERROR.getText()));
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCopybookWithIndirectRecursiveDependencyIsDetected.java:59]: in lsp4j  Diagnostic constructor defined with @NonNull, is that validation strict enough or ignored?  Same for all constructor where "null" used.

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestEntireProgramInCopybook.java [hunk 6]
```diff
@@ -16,109 +16,37 @@
...
-
-  private void assertVariableUsages(Map<String, List<Location>> usages) {
-    assertEquals("Variable usages: " + usages.toString(), 1, usages.size());
-
-    assertNumberOfLocations(usages, CHILD1, 1);
-    assertLocation(usages, CHILD1, PROGA_URI, 7, 19);
+  public void test() {
+    UseCaseEngine.runTest(
+        TEXT,
+        asList(new CobolText(PROGA_NAME, PROGA), new CobolText(STRUCT1_NAME, STRUCT1)),
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestEntireProgramInCopybook.java:49]: If list usage immutable then better use List.of( .... )

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestEntireProgramInCopybook.java [hunk 7]
```diff
@@ -16,109 +16,37 @@
...
+        emptyMap());
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestEntireProgramInCopybook.java:50]: Map.of()   (for all occurrences)

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCobolStringWithContinuationLineConcatenatedCorrectly.java [hunk 8]
```diff
@@ -13,32 +13,42 @@
...
+          + "000000     END-EVALUATE.\n"
+          + "000000 {#*PROGRAM2}.\n"
+          + "000000  DISPLAY 'HELLO'.";
 
+  @Ignore("Continuation line cause a semantic error #414")
+  // TODO: #414
   @Test
   public void test() {
-    super.test();
+    UseCaseEngine.runTest(TEXT, emptyList(), emptyMap());
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCobolStringWithContinuationLineConcatenatedCorrectly.java:52]: List.of(), Map.of(),  immutable empty

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/TestPreprocessorListenerImpl.java [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+  }
+
+  @Override
+  public void exitVariableStatement(VariableStatementContext ctx) {
+    pop();
+    ofNullable(ctx.variableUsage())
+        .map(VariableUsageContext::word)
+        .ifPresent(
+            it ->
+                processToken(
```
Comment (grianbrcom) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/TestPreprocessorListenerImpl.java:151]: I think you could overload `processToken` like: 

```processToken(WordContext wc, Map<String, List<Location>> theMap, VariableStatementContext ctx):```

And inside this function extract text, replacement and diagnostic.

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestExtraSymbolsNotCauseErrorOnNextLine.java [hunk 10]
```diff
@@ -13,43 +13,33 @@
...
-
-    // The position of dot at the end of line "INITIALIZE ID4(ROW-SUB)."
-    Range range = diagnostic.getRange();
-    assertEquals("Diagnostic start line", 0, range.getStart().getLine());
-    assertEquals("Diagnostic start character", 80, range.getStart().getCharacter());
-    assertEquals("Diagnostic end line", 0, range.getEnd().getLine());
-    assertEquals("Diagnostic end character", 86, range.getEnd().getCharacter());
+    UseCaseEngine.runTest(
+        TEXT,
+        emptyList(),
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestExtraSymbolsNotCauseErrorOnNextLine.java:41]: List.of()    (for all occurrences)

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestExtraneousInputEOFExpecting.java [hunk 11]
```diff
@@ -13,52 +13,41 @@
...
-        diagnostic.getMessage());
-
-    Range range = diagnostic.getRange();
-    assertEquals("Diagnostic start line", 5, range.getStart().getLine());
-    assertEquals("Diagnostic start character", 35, range.getStart().getCharacter());
-    assertEquals("Diagnostic end line", 5, range.getEnd().getLine());
-    assertEquals("Diagnostic end character", 35, range.getEnd().getCharacter());
+    UseCaseEngine.runTest(
+        TEXT,
+        emptyList(),
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestExtraneousInputEOFExpecting.java:49]: List.of()

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestIdentificationDivisionInCopybook.java [hunk 12]
```diff
@@ -16,112 +16,39 @@
...
-  private void assertVariableUsages(Map<String, List<Location>> usages) {
-    assertEquals("Variable usages: " + usages.toString(), 1, usages.size());
-
-    assertNumberOfLocations(usages, CHILD1, 1);
-    assertLocation(usages, CHILD1, DOCUMENT_URI, 7, 19);
+  public void test() {
+    UseCaseEngine.runTest(
+        TEXT,
+        asList(new CobolText(IDDIV_NAME, IDDIV), new CobolText(STRUCT1_NAME, STRUCT1)),
+        emptyMap());
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestIdentificationDivisionInCopybook.java:52]: Map.of()

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestIdentificationDivisionInCopybook.java [hunk 13]
```diff
@@ -16,112 +16,39 @@
...
-
-  private void assertVariableUsages(Map<String, List<Location>> usages) {
-    assertEquals("Variable usages: " + usages.toString(), 1, usages.size());
-
-    assertNumberOfLocations(usages, CHILD1, 1);
-    assertLocation(usages, CHILD1, DOCUMENT_URI, 7, 19);
+  public void test() {
+    UseCaseEngine.runTest(
+        TEXT,
+        asList(new CobolText(IDDIV_NAME, IDDIV), new CobolText(STRUCT1_NAME, STRUCT1)),
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestIdentificationDivisionInCopybook.java:51]: List.of(... ) if immutable usage,   (for all occurrences)

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestMappingWithMultiDefinitions.java [hunk 14]
```diff
@@ -95,151 +89,15 @@
...
-    assertNumberOfLocations(variableDefinitions, "CHILD2-NESTED", 1);
-    assertLocation(variableDefinitions, "CHILD2-NESTED", toURI(NESTED1_NAME), 2, 15);
-
-    assertNumberOfLocations(variableDefinitions, "CHILD3-NESTED", 1);
-    assertLocation(variableDefinitions, "CHILD3-NESTED", toURI(NESTED2_NAME), 0, 15);
-    assertNumberOfLocations(variableDefinitions, "CHILD4-NESTED", 1);
-    assertLocation(variableDefinitions, "CHILD4-NESTED", toURI(NESTED2_NAME), 1, 15);
+    UseCaseEngine.runTest(
+        DOCUMENT,
+        asList(
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestMappingWithMultiDefinitions.java:94]: same

**Comment by temanbrcom** [comment]
It is generated, but I decided to commit it as is because I don't want to set up generating on the build - it won't go to production anyway. Also, it is not something that is going to change frequently.

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/UseCaseEngine.java [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+        (key, value) ->
+            assertEquals(
+                message,
+                value.stream().sorted(getLocationComparator()).collect(toList()),
+                actual.get(key).stream().sorted(getLocationComparator()).collect(toList())));
+  }
+
+  private Comparator<Location> getLocationComparator() {
+    return comparing(Location::getUri)
+        .thenComparingInt(location -> location.getRange().getStart().getLine())
```
Comment (asatklichov) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/UseCaseEngine.java:216]: 👍

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCobolStringWithContinuationLineConcatenatedCorrectly.java (hunk 8 above)
Comment (ishche) [reply com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCobolStringWithContinuationLineConcatenatedCorrectly.java:52]: https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#emptyList--

emptyList() returns immutable list too

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCopybookWithIndirectRecursiveDependencyIsDetected.java (hunk 5 above)
Comment (temanbrcom) [reply com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCopybookWithIndirectRecursiveDependencyIsDetected.java:59]: If I understand correctly, this validation applies only while serialization to JSON.
Technically, this usage is not correct, but this check won't be applied in this scenario.

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCobolStringWithContinuationLineConcatenatedCorrectly.java (hunk 8 above)
Comment (asatklichov) [reply com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCobolStringWithContinuationLineConcatenatedCorrectly.java:52]: Yes, but List.of() is much more safer, e.g. you can deserialize emptyList() on JDK 8 and previous, but not List.of.

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/TestPreprocessorListenerImpl.java (hunk 9 above)
Comment (temanbrcom) [reply com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/TestPreprocessorListenerImpl.java:151]: Not exactly - it should consume different types of statement contexts, e.g. `ParagparphStatementContext` or `ErrorStatementContext`. They have a quite similar structure, but the limitations of the Java object model won't allow using it this way

Comment (grianbrcom) [reply com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/TestPreprocessorListenerImpl.java:151]: Yes, I see.

**Pull request merged**